### DISCUSSION TITLE
collection/arraylist

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#95] [Announcements]
# collection/arraylist

Java程序员进阶之路，小白的零基础Java教程，Java ArrayList详解

https://tobebetterjavaer.com/collection/arraylist.html

**Comment by 2565338472** [comment]
提个建议，最好文档可以标注好jdk 的版本，所以请问这里的源码使用的是那个版本的jdk

**Reply by itwanger** [reply]
如非特殊说明，一般都会用 JDK 8 这难道不是大家的共识吗，😝

**Comment by haolovejin** [comment]
针对：

“初始化的时候指定一个大小可以有效地避免在添加新元素的时候进行不必要的扩容。
List<String> alist = new ArrayList<>(20);“

这一内容。我个人在学习的时候有疑惑，不是很明白为什么，后来查阅资料，特此分享一下我的理解：

ArrayList会根据需要自动增加内部数组的大小以容纳更多的元素，通常扩容策略是将当前数组容量扩大一倍。
如果没有指定初始大小，可能一开始的数组容量是10？（这里我也不确定具体是多少），反正一开始的数组容量是比较小的，如果没有初始化的大小就可能会出现频繁扩容，扩容操作涉及创建新的数组，然后复制原有数组到新数组中，这会产生一定的性能开销，所以初始化大小可以提高效率。

**Comment by icc0216** [comment]
private void ensureCapacityInternal(int minCapacity) {
        ensureExplicitCapacity(calculateCapacity(elementData, minCapacity));
    }

ArrayList的这个方法源码改了
